packages/regl-worldview/src/commands/DrawPolygon/PolygonBuilder.js

```diff
@@ -0,0 +1,380 @@
+// @flow
+
+//  Copyright (c) 2019-present, GM Cruise LLC
+//
+//  This source code is licensed under the Apache License, Version 2.0,
+//  found in the LICENSE file in the root directory of this source tree.
+//  You may not use this file except in compliance with the License.
+
+import distance from "distance-to-line-segment";
+
+import type { Vec3, MouseHandler } from "../../types";
+import { PolygonPoint, Polygon } from "./index";
+
+type OptionalZPoint = {
+  x: number,
+  y: number,
+  z?: number,
+};
+
+type PolygonAddCommand = {
+  points: OptionalZPoint[],
+  name?: string,
+};
+
+function areEqual(point1: PolygonPoint, point2: PolygonPoint) {
+  const [x1, y1, z1] = point1.point;
+  const [x2, y2, z2] = point2.point;
+  return x1 === x2 && y1 === y2 && z1 === z2;
+}
+
+function isClosed(polygon: Polygon): boolean {
+  const { points } = polygon;
+  for (let i = 0; i < points.length - 1; i++) {
+    if (areEqual(points[i], points[i + 1])) {
+      return true;
+    }
+  }
+  const firstPoint = points[0];
+  const lastPoint = points[points.length - 1];
+  return areEqual(firstPoint, lastPoint);
+}
+
+// has listeners you can pass to Worldview for mouse interactions
+// internally builds a list of polygons and modifies the polygons
+// based on mouse & keyboard interactions.  for now we use mututation internally
+// instead of immutability to keep the number of allocations lower and make
+// the implementation a bit more straightforward
+//
+// for an example implementation check stories/Worldview/drawPolygon.js
+export default class PolygonBuilder {
+  mouseDown: boolean = false;
+  polygons: Polygon[];
+  onChange: () => void = () => {};
+  activePolygon: ?Polygon;
+  activePoint: ?PolygonPoint;
+  mouseDownPoint: Vec3;
```

Comment:
Might be good to prefix private variables and methods with an underscore.

Reply:
`mouseDownPoint` and other instance variable seem to be useful for public api as well, e.g. trigger certain external events based mouseDown position